scapy/contrib/bgp.py

```diff
@@ -8,162 +8,629 @@
 from scapy.packet import *
 from scapy.fields import *
 from scapy.layers.inet import TCP
+from scapy.contrib.bgp_fields import *
 
+def classDispatcher(s,classDict,defClass,index_from=None):
+    """ A generic class dispatcher:
+    s:          packet
+    classDict:  a dictionary of index:packetClass
+    defClass:   a default packet class when index is not found
+    index_from: normally a lambda s: to calculate the index for the classDict
 
-class BGPIPField(Field):
-	"""Represents how bgp dose an ip prefix in (length, prefix)"""
-	def mask2iplen(self,mask):
-		"""turn the mask into the length in bytes of the ip field"""
-		return (mask + 7) // 8
-	def h2i(self, pkt, h):
-		"""human x.x.x.x/y to internal"""
-		ip,mask = re.split( '/', h)
-		return  int(mask), ip
-	def i2h( self, pkt, i):
-		mask, ip = i
-		return ip + '/' + str( mask )
-	def i2repr( self, pkt, i):
-		"""make it look nice"""
-		return self.i2h(pkt,i)
-	def i2len(self, pkt, i):
-		"""rely on integer division"""
-		mask, ip = i
-		return self.mask2iplen(mask) + 1
-	def i2m(self, pkt, i):
-		"""internal (ip as bytes, mask as int) to machine"""
-		mask, ip = i
-		ip = inet_aton( ip )
-		return struct.pack(">B",mask) + ip[:self.mask2iplen(mask)] 
-	def addfield(self, pkt, s, val):
-		return s+self.i2m(pkt, val)
-	def getfield(self, pkt, s):
-		l = self.mask2iplen( struct.unpack(">B",s[0])[0] ) + 1
-		return s[l:], self.m2i(pkt,s[:l])
-	def m2i(self,pkt,m):
-		mask = struct.unpack(">B",m[0])[0]
-		ip = "".join(m[i + 1] if i < self.mask2iplen(mask) else '\x00'
-                     for i in xrange(4))
-		return (mask,inet_ntoa(ip))
+    return Raw(s) if index_from throws an Exception"""
+    try:
```

Comment:
Rhabks for the review. It seems I have homework
;-)
El 29/1/2016 13:43, "Pierre Lalet" <notifications@github.com> escribió:

> In scapy/contrib/bgp.py
> <https://github.com/secdev/scapy/pull/10#discussion_r51256510>:
>
> > -		"""internal (ip as bytes, mask as int) to machine"""
> > -		mask, ip = i
> > -		ip = inet_aton( ip )
> > -		return struct.pack(">B",mask) + ip[:self.mask2iplen(mask)]
> > -	def addfield(self, pkt, s, val):
> > -		return s+self.i2m(pkt, val)
> > -	def getfield(self, pkt, s):
> > -		l = self.mask2iplen( struct.unpack(">B",s[0])[0] ) + 1
> > -		return s[l:], self.m2i(pkt,s[:l])
> > -	def m2i(self,pkt,m):
> > -		mask = struct.unpack(">B",m[0])[0]
> > -		ip = "".join(m[i + 1] if i < self.mask2iplen(mask) else '\x00'
> > -                     for i in xrange(4))
> > -		return (mask,inet_ntoa(ip))
> > +    return Raw(s) if index_from throws an Exception"""
> > +    try:
>
> I believe returning Raw should be the default when conf.debug_dissector
> is set to 0 / False. See #40 <https://github.com/secdev/scapy/pull/40>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/secdev/scapy/pull/10/files#r51256510>.
>